PSPs/drafts/psp-34.md

```diff
@@ -0,0 +1,914 @@
+This proposal aims to resolve that by defining one **interface** that shares the same **ABI** of **permissionless** methods between all implementations.
+
+The goal is to have a standard contract interface that allows tokens deployed on Substrate's `contracts` pallet to be re-used by other applications: from wallets to decentralized exchanges.
+
+## Implementations
+Examples of implementations:
+
+- [OpenBrush](https://github.com/Supercolony-net/openbrush-contracts/blob/main/contracts/token/psp721/psp721.rs), written in [ink!](https://github.com/paritytech/ink).
+
+## Motivation for having a standard separate from ERC-721
+Due to the different nature of WebAssembly smart contracts and the difference between EVM and the [`contracts` pallet](https://github.com/paritytech/substrate/tree/master/frame/contracts) in Substrate, this standard proposal has specific rules and methods,
+therefore PSP-34 differs from ERC-721 in its implementation. Also the proposal contains new methods that should improve the usability of Non-Fungible Token.
+
+# This standard is at ABI level
+
+Substrate's [`contracts` pallet](https://github.com/paritytech/substrate/tree/master/frame/contracts) can execute any WebAssembly contract that implements its defined API; we do not want to restrain this standard to only Rust and [the ink! language](https://github.com/paritytech/ink), but make it possible to be implemented by any language/framework that compiles to WebAssembly.
+
+## Specification
+1. [Interfaces](#Interfaces)
+2. [Extension](#Extension)
+3. [Events](#Events)
+4. [Types](#Types)
+5. [Errors](#Errors)
+
+### Interfaces
+
+#### PSP-34 Interface
+
+This section defines the required interface for this standard.
+
+##### **collection_id**() ➔ Id
+Selector: `0xffa27a5f` - first 4 bytes of `blake2b_256("PSP34::collection_id")`
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the collection `Id` of the NFT token.",
+    "",
+    "This can represents the relationship between tokens/contracts/pallets."
+  ],
+  "mutates": false,
+  "label": "PSP34::collection_id",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Id"
+    ],
+    "type": "Id"
+  },
+  "selector": "0xffa27a5f"
+}
+```
+
+##### **balance_of**(owner: AccountId) ➔ u32
+Selector: `0xcde7e55f` - first 4 bytes of `blake2b_256("PSP34::balance_of")`
+```json
+{
+  "args": [
+    {
+      "label": "owner",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    }
+  ],
+  "docs": [
+    "Returns the balance of the owner.",
+    "",
+    "This represents the amount of unique tokens the owner has."
+  ],
+  "mutates": false,
+  "label": "PSP34::balance_of",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "u32"
+    ],
+    "type": 1
+  },
+  "selector": "0xcde7e55f"
+}
+```
+
+##### **owner_of**(id: Id) ➔ AccountId
```

Comment:
Does it sound ridiculous if we return `Option<AccountId>`?

Reply:
I think it is not a bad idea to return `Option` in case if owner or id is not exist

Reply:
```suggestion
##### **owner_of**(id: Id) ➔ Option<AccountId>
```